scripts/benchmark_db/benchmark_base.rb

```diff
@@ -0,0 +1,263 @@
+    else
+      @logger.info 'Cloning repository'
+      clone_command
+      Dir.mkdir(repository_name) if @dry_run
+      @created_repository = true
+    end
+
+    @logger.info 'Clean and build client'
+    Dir.chdir(repository_name) do
+      checkout_command
+      clean_command
+      build_command
+    end
+  end
+
+  # Calls all helper functions to prepare each client to run benchmarks.
+  def build_artefacts
+    @logger.info 'Building artefacts...'
+    build_client
+
+    build_config_file unless @dry_run
+    build_substitution_hash
+  end
+  private :build_artefacts
+end
+
+# Mixin module for base benchmark class run and validation commands.
+module RunCommands
+  # Create and call proper validation command, then write results to metrics.
+  def run_validation_step(daily, args, metrics)
+    unless daily
+      validate_command = splice_args(@validate_command, args)
+      metrics[:validate] = exec_command(validate_command)
+      return
+    end
+
+    validate_online_command = splice_args(@validate_online_command, args)
+    new_metrics = exec_command(validate_online_command, self)
+    new_metrics[:tpm] =
+      new_metrics[:num_epochs] ? new_metrics[:num_epochs] / online_validation_secs : 'n/a'
+    new_metrics[:tpm] = new_metrics[:tpm].ceil(3)
+    metrics[:validate_online] = new_metrics
+  end
+
+  # Import snapshot, write metrics, and call validation function, returning metrics.
+  def import_and_validation(daily, args, metrics)
+    import_command = splice_args(@import_command, args)
+    metrics[:import] = exec_command(import_command)
+
+    # Save db size just after import.
+    metrics[:import][:db_size] = db_size unless @dry_run
+
+    run_validation_step(daily, args, metrics)
+    metrics
+  rescue StandardError, Interrupt
+    @logger.error('Fiasco during benchmark run. Deleting downloaded files, cleaning DB and stopping process...')
+    FileUtils.rm_f(@snapshot_path) if @snapshot_downloaded
+    clean_db
+    FileUtils.rm_rf(repository_name) if @created_repository
+    exit(1)
+  end
+
+  # This is the primary function called in `bench.rb` to run the metrics for
+  # each benchmark.
+  def run(daily, snapshot_downloaded)
+    begin
+      @snapshot_downloaded = snapshot_downloaded
+      @logger.info "Running bench: #{@name}"
+
+      metrics = Concurrent::Hash.new
+      args = build_artefacts
+      @sync_status_command = splice_args(@sync_status_command, args)
+
+      exec_command(@init_command) if @name == 'forest'
+
+      @metrics = import_and_validation(daily, args, metrics)
+    rescue StandardError, Interrupt
+      @logger.error('Fiasco during benchmark run. Deleting downloaded files and stopping process...')
+      FileUtils.rm_f(@snapshot_path) if @snapshot_downloaded
+      FileUtils.rm_rf(repository_name) if @created_repository
+      exit(1)
+    end
+
+    @logger.info 'Cleaning database'
+    clean_db
+    @logger.info 'Deleting downloaded repository'
+    delete_repository
```

Comment:
Why do we have a `delete_repository` here?

We don't want to delete git repo between runs, otherwise a benchmark that will test different configurations of forest (i.e. paritydb options) will have to clone the repository each time, making the whole benchmark process slower.

Reply:
As we discussed, adding this to #2683